thewalrus/reference.py

```diff
@@ -278,3 +282,121 @@ def hafnian(M, loop=False):
+    Blist = [Blist[0][i] for i in tuple((0,) + perm)] + [Blist[1][i] for i in tuple((0,) + perm)]
+    dico_list = {j: i + 1 for i, j in enumerate(Blist)}
+    new_mapping_list = {
+        objects[dico_list[i] - 1]: objects[dico_list[j] - 1]
+        for i, j in zip(list(range(0, m - 1)) + [m], list(range(m + 1, 2 * m)) + [m - 1])
+    }
+    return tuple(new_mapping_list.items())
+
+
+def bitstrings(n):
+    """Returns the bistrings from 0 to n/2
+
+    Args:
+        n (int): Twice the highest bitstring value.
+
+    Returns:
+        (iterator): An iterable of all bistrings.
+    """
+    for binary in map("".join, product("01", repeat=n - 1)):
+        yield "0" + binary
+
+
+def rpmp(s):
+    """Generates the restricted set of perfect matchings matching permutations.
+
+    Args:
+        s (tuple): tuple of labels to be used
+
+    Returns:
+        generator: the set of restricted perfect matching permutations of the tuple ``s``
+    """
+    m = len(s) // 2
+
+    def local_mapper(x):
+        """Helper function to define a local mapper based on the symbols s
+        Args:
+            x (iterable): object to be mapped
+        """
+        return mapper(x, s)
+
+    for i in product(permutations(range(1, m)), bitstrings(m)):
+        yield local_mapper(i)
+
+
+def splitter(elem):
+    """Takes an element from the restricted perfect matching permutations (rpmp) and returns all the associated elements in the restricted single pair matchings (rspm)
+
+    Args:
+        elem (tuple): tuple representing an element of rpmp
+
+    Returns:
+        (iterator): all the associated elements in rspm
+    """
+    num_elem = len(elem)
+    net = [elem]
+    for i in range(num_elem):
+        left = (elem[j] for j in range(i))
+        middle_left = ((elem[i][0], elem[i][0]),)
+        middle_right = ((elem[i][1], elem[i][1]),)
+        right = (elem[j] for j in range(i + 1, num_elem))
+        net.append(tuple(middle_right) + tuple(right) + tuple(left) + tuple(middle_left))
+    for i in net:
+        yield i
+
+
+def rspm(s):
+    """Generates the restricted set of single-pair matchings.
+
+    Args:
+        s (tuple): tuple of labels to be used
+
+    Returns:
+        generator: the set of restricted perfect matching permutations of the tuple s
+    """
+    gen = rpmp(s)
+    return chain(*(splitter(i) for i in gen))
+
+
+def mtl(A, loop=False):
+    """Returns the Montrealer of an NxN matrix and an N-length vector.
+
+    Args:
+        A (array): an NxN array of even dimensions. Can be symbolic.
+        loop (boolean): if set to ``True``, the loop montrealer is returned
+
+    Returns:
+        np.float64, np.complex128 or sympy.core.add.Add: the Montrealer of matrix ``A``.
+    """
+    n, _ = A.shape
+    net_sum = 0
+
+    perm = rspm(range(n)) if loop else rpmp(range(n))
+    for s in perm:
+        net_prod = 1
+        for a in s:
+            net_prod *= A[a[0], a[1]]
+
+        net_sum += net_prod
+
+    return net_sum
```

Comment:
This could be numbified if the matrix A is not symbolic.

Reply:
Now I understood this is the reference implementation so it is not critical to separate the concerns between the numerical and symbolical implementation